### PR TITLE
Feature/tools

### PR DESCRIPTION
**Pull request by justliliandev**
# Description

Added Yeta Wrench:
- Changed the rotation to only allow for valid blockstates
Added the Coordination Selector and Location Printout

# Todo 

- [x] Lang Datagen
- [x] ItemStack Capability
- [x] PR to Art-Repo to fix texture rename and missing gui texture

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by justliliandev** [comment]
I've also renamed renderers to renderer because package names are in singular form

File: enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        font.drawShadow(pPoseStack, txt, x, y, 0xFFFFFF);
+
+        txt = ICoordinateSelection.getLevelName(getMenu().getSelection().getLevel());
+        y += font.lineHeight + 4;
+        x = midX - font.width(txt) / 2;
+        font.drawShadow(pPoseStack, txt, x, y, 0xFFFFFF);
+    }
+
+    @Override
+    protected void renderLabels(PoseStack pPoseStack, int pMouseX, int pMouseY) {
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java]: Can we maybe move some common code ino an EIOScreen class or something? I feel like this override and a load others are going to be pretty common...

File: enderio-base/src/main/java/com/enderio/base/common/menu/CoordinateMenu.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        this.selection = selection;
+        this.isPrintout = name != null;
+        this.name = name != null ? name : "";
+    }
+
+    public static CoordinateMenu factory(@Nullable MenuType<CoordinateMenu> pMenuType, int pContainerId, Inventory inventory, FriendlyByteBuf buf) {
+        return new CoordinateMenu(pMenuType, pContainerId, inventory, buf);
+    }
+
+        /**
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/menu/CoordinateMenu.java]: Formatting error :)

File: enderio-base/src/main/java/com/enderio/base/common/menu/CoordinateMenu.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        buf.writeUtf(name == null ? "" : name, 50);
+        return buf;
+    }
+
+    @Override
+    public boolean stillValid(Player pPlayer) {
+        return findInHand(pPlayer, isPrintout ? EIOItems.LOCATION_PRINTOUT.get() : EIOItems.COORDINATE_SELECTOR.get());
+    }
+
+
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/menu/CoordinateMenu.java]: Blank space :)

File: enderio-base/src/main/java/com/enderio/base/common/network/EIOPackets.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import net.minecraftforge.fmllegacy.network.NetworkRegistry;
+import net.minecraftforge.fmllegacy.network.simple.SimpleChannel;
+
+public class EIOPackets {
+
+    private EIOPackets() {}
+
+    private static final String PROTOCOL_VERSION = "1.0";
+    public static final SimpleChannel INSTANCE = NetworkRegistry.newSimpleChannel(
+        new ResourceLocation(EnderIO.DOMAIN, "network"), //TODO: Enderio.rl
```
Comment (Rover656) [inline enderio-base/src/main/java/com/enderio/base/common/network/EIOPackets.java]: TODO now invalid, you can change this

**Review by Rover656 — CHANGES_REQUESTED** [review]

File: enderio-base/src/main/java/com/enderio/base/common/capability/location/ICoordinateSelection.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    default void copyInto(ICoordinateSelection other) {
+        setLevel(other.getLevel());
+        setPos(other.getPos());
+        set(other.isSet());
+    }
+
+    boolean isSet();
+
+    void set(boolean isSet);
```
Comment (CrazyPants) [inline enderio-base/src/main/java/com/enderio/base/common/capability/location/ICoordinateSelection.java]: setSet, or maybe setIsSet

File: enderio-base/src/main/java/com/enderio/base/common/item/tool/CoordinateSelectorItem.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    public InteractionResult useOn(UseOnContext pContext) {
+        if (pContext.getPlayer() != null && checkPaper(pContext.getPlayer())) {
+            if (pContext.getPlayer() instanceof ServerPlayer serverPlayer)
+                openMenu(serverPlayer, pContext.getClickedPos(), pContext.getLevel().dimension().location());
+            return InteractionResult.sidedSuccess(pContext.getLevel().isClientSide);
+        }
+        return super.useOn(pContext);
+    }
+
+    public static void openMenu(ServerPlayer player, BlockPos pos, ResourceLocation dimension) {
```
Comment (CrazyPants) [inline enderio-base/src/main/java/com/enderio/base/common/item/tool/CoordinateSelectorItem.java]: this looks like it should be a private non-static method

File: enderio-base/src/main/java/com/enderio/base/common/menu/CoordinateMenu.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.inventory.MenuType;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import org.apache.commons.lang3.StringUtils;
+
```
Comment (CrazyPants) [inline enderio-base/src/main/java/com/enderio/base/common/menu/CoordinateMenu.java:19]: Some unused imports

File: enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
+package com.enderio.base.client.screen;
+
+import com.enderio.base.EnderIO;
+import com.enderio.base.common.capability.location.ICoordinateSelection;
+import com.enderio.base.common.menu.CoordinateMenu;
+import com.enderio.base.common.network.EIOPackets;
+import com.enderio.base.common.network.packet.UpdateCoordinateSelectionNameMenuPacket;
+import com.enderio.base.common.capability.location.CoordinateSelection;
```
Comment (CrazyPants) [inline enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java]: unused import

File: enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java (hunk 1 above)
Comment (justliliandev) [reply enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java]: I created a blank EIOScreen in EnderCore with EditBoxHelpers for all screens

Comment (Rover656) [reply enderio-base/src/main/java/com/enderio/base/client/screen/CoordinateMenuScreen.java]: Looking good :)

**Review by Rover656 — APPROVED** [review]

**Comment by Rover656** [comment]
All good for merge if we're all happy :)

**Comment by Rover656** [comment]
Uh actually, GH Actions failed; mind having a look?

**Review by CrazyPants — APPROVED** [review]

**Comment by Rover656** [comment]
Nice work; thanks :)

**Pull request merged**